README.md

### Extra Notes

Check out:
  - https://nurbs-python.readthedocs.io/en/5.x/index.html
    - Relatively complete. Includes trim curves but not intersection
      computations.
  - https://blog.mapbox.com/drawing-antialiased-lines-with-opengl-8766f34192dc
    - May be useful for WGPU / OpenGL drawing of tessellated NURBS.


CHANGELOG.md
# Revision history for transom

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Test.hs
module Main (main) where

import Hedgehog (Group, checkParallel)
import qualified REYES.SamplerTest (tests)
import Test.DocTest (doctest)

main :: IO ()
main = do
  runHedgehogTests
  runDocTests

runHedgehogTests :: IO ()
runHedgehogTests = do
  putStrLn "\n---- Running Hedgehog Tests ----"
  mapM_ checkParallel hedgehogTests
  putStrLn "---- Finished Hedgehog Tests ----"

hedgehogTests :: [Group]
hedgehogTests =
  [REYES.SamplerTest.tests]

runDocTests :: IO ()
runDocTests = do
  putStrLn "\n---- Running Doctests ----"
  docTests
  putStrLn "---- Finished Doctests ----"

docTests :: IO ()
docTests =
  doctest
    [ "-isrc",
      "src/REYES/Sampler.hs"
    ]


test/cbits/kensler.c
/** Kensler (2013)
  *
  * This file contains functions that are (slightly modified) from:
  *
  *   * Kensler, A (2013) Correlated Multi-Jittered Sampling. Pixar Technical
  *     Memo, 13-01.
  */

#include <stdint.h>

/** Permute function from Listing 3.
  *
  * @param i index into the permutation [0, l-1]
  * @param l length of the permutation
  * @param p permutation number (any value)
  */
unsigned permute(uint32_t i, uint32_t l, uint32_t p) {
  uint32_t w = l - 1;
  w |= w >> 1;
  w |= w >> 2;
  w |= w >> 4;
  w |= w >> 8;
  w |= w >> 16;
  do {
    i ^= p;              i *= 0xe170893d;
    i ^= p       >> 16;
    i ^= (i & w) >>  4;
    i ^= p       >>  8;  i *= 0x0929eb3f;
    i ^= p       >> 23;
    i ^= (i & w) >>  1;  i *= 1 | p >> 27;
                         i *= 0x6935fa69;
    i ^= (i & w) >> 11;  i *= 0x74dcb303;
    i ^= (i & w) >>  2;  i *= 0x9e501cc3;
    i ^= (i & w) >>  2;  i *= 0xc860a3df;
    i &= w;
    i ^= i       >>  5;
  } while (i >= l);
  return (i + p) % l;
}


test/REYES/SamplerTest.hs
{-# LANGUAGE OverloadedStrings #-}

module REYES.SamplerTest (tests) where

import Control.Monad.IO.Class (liftIO)
import Data.List (sort)
import Data.Word (Word32)
import Hedgehog (Gen, Group (Group), Property, forAll, property, (===))
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified REYES.Sampler as Sampler

tests :: Group
tests =
  Group
    "REYES.Sampler"
    [ ("permute matches C implementation", prop_permute_vs_C),
      ("permute must be complete", prop_permute_complete)
    ]

-- | All values in a 'Sampler.permute' sequence must be present.
prop_permute_complete :: Property
prop_permute_complete = property $ do
  l <- forAll $ Gen.integral (Range.linear 1 (128 ^ 2))
  p <- forAll $ Gen.integral (Range.linear minBound maxBound)

  let orderedSeq :: [Word32]
      orderedSeq = [0 .. (l - 1)]

      permutedSeq :: [Word32]
      permutedSeq = (\i -> Sampler.permute i l p) <$> orderedSeq

      sortedPermutation :: [Word32]
      sortedPermutation = sort permutedSeq

  sortedPermutation === orderedSeq

-- | Check that the 'Sampler.permute' function matches a C implementation.
--
-- The C implementation is from Listing 3 of:
--
--   * Kensler, A (2013) Correlated Multi-Jittered Sampling. Pixar Technical
--     Memo, 13-01.
prop_permute_vs_C :: Property
prop_permute_vs_C = property $ do
  l <- forAll $ Gen.integral (Range.linear 1 maxBound)
  i <- forAll $ Gen.integral (Range.linear 0 (l - 1))
  p <- forAll $ Gen.integral (Range.linear minBound maxBound)
  result_c <- liftIO $ permute_c i l p
  let result_hs :: Word32
      result_hs = Sampler.permute i l p
  result_hs === result_c

-- | The original @permute@ function from Kensler (2013):
--
--   * Kensler, A (2013) Correlated Multi-Jittered Sampling. Pixar Technical
--     Memo, 13-01.
foreign import ccall "permute"
  permute_c :: Word32 -> Word32 -> Word32 -> IO Word32


src/REYES/Sampler.hs
{-# LANGUAGE BangPatterns #-}

module REYES.Sampler where

import Data.Bits (shiftR, xor, (.&.), (.|.))
import Data.Word (Word32)

-- | Pseudorandom permutation function.
--
-- The permutation function @'permute' index size pat@ chooses a random integer
-- from the range @[0, size-1]@. @index@ specifies which integer to choose, such
-- that each integer is chosen only once. @pat@ specifies which random pattern
-- to draw from, such that each value of @pat@ should produce a different random
-- sequence.
--
-- For example, we may compute a full random permutation of the sequence
-- @[0,1,2,3,4]@ by drawing the permutation at each index as follows:
--
-- >>> let size = 5  -- 5 elements, 0 to 4
-- >>> let pat = 0
-- >>> (\i -> permute i size pat) <$> [0 .. size-1]
-- [0,4,2,3,1]
--
-- Each @pat@ value selects a different permutation pattern:
--
-- >>> let pat = 1
-- >>> (\i -> permute i size pat) <$> [0 .. size-1]
-- [0,1,2,4,3]
--
-- This is Listing 3 from Kensler, 2013:
--
--   * Kensler, A (2013) Correlated Multi-Jittered Sampling. Pixar Technical
--     Memo, 13-01.
permute ::
  -- | Index into the permutation. This must be a value in the range
  --   @[0, size-1]@.
  Word32 ->
  -- | Length or size of the permutation. How many elements it has.
  Word32 ->
  -- | Pattern number. The instance of the permutation from which we are
  --   drawing elements.
  Word32 ->
  -- | Permuted value.
  Word32
permute !index !size !pat =
  let orEq :: (Word32 -> Word32) -> Word32 -> Word32
      orEq !f !x = x .|. f x

      xorEq :: (Word32 -> Word32) -> Word32 -> Word32
      xorEq !f !x = x `xor` f x

      -- flip (.) so that operations read in the same direction as the original
      -- C code
      (.>) :: (a -> b) -> (b -> c) -> a -> c
      (.>) = flip (.)

      w :: Word32
      w =
        orEq (`shiftR` 1)
          .> orEq (`shiftR` 2)
          .> orEq (`shiftR` 4)
          .> orEq (`shiftR` 8)
          .> orEq (`shiftR` 16)
          $ size - 1

      cycleWalk :: Word32 -> Word32
      cycleWalk !q =
        let r :: Word32
            r =
              xor pat
                .> (*) 0xe170893d
                .> xor (pat `shiftR` 16)
                .> xorEq ((.&.) w .> (`shiftR` 4))
                .> xor (pat `shiftR` 8)
                .> (*) 0x0929eb3f
                .> xor (pat `shiftR` 23)
                .> xorEq ((.&.) w .> (`shiftR` 1))
                .> (*) (1 .|. (pat `shiftR` 27))
                .> (*) 0x6935fa69
                .> xorEq ((.&.) w .> (`shiftR` 11))
                .> (*) 0x74dcb303
                .> xorEq ((.&.) w .> (`shiftR` 2))
                .> (*) 0x9e501cc3
                .> xorEq ((.&.) w .> (`shiftR` 2))
                .> (*) 0xc860a3df
                .> (.&.) w
                .> xorEq (`shiftR` 5)
                $ q
         in if r >= size then cycleWalk r else r
   in (cycleWalk index + pat) `mod` size
